### ISSUE TITLE
Perplexity (ppl) Calculation of Local Sparse Model: NaN issue

### ISSUE THREAD

**Issue by HengJayWang** [vllm-project/llm-compressor#853]
👋 Hello Neural Magic community developers,

I encountered an issue while calculating the perplexity for a locally converted Llama3-8B sparse model using the llm-compress library. I'm refer the sparse conversion [example script](https://github.com/vllm-project/llm-compressor/blob/main/examples/quantization_24_sparse_w4a16/llama7b_sparse_w4a16.py) and change model to [meta-llama/Meta-Llama-3-8B-Instruct](https://huggingface.co/meta-llama/Meta-Llama-3-8B-Instruct) by my self, the sparse conversion need ~ 1.2 hours to finish. 
Here’s a detailed breakdown:

**Describe the bug**
While trying to compute the WikiText2 Perplexity for a Llama3-8B model that has been sparsified (load local model from disk), the resulting perplexity values always turn out to be NaN. I suspect that some configurations might not be properly set when using the custom SparseAutoModelForCausalLM class in combination with the compressed-tensors library.

**Expected behavior**
I expected the perplexity values to be reasonable and comparable to the official Hugging Face models. For example, when testing with the standard Llama-3.2-3B model from Hugging Face (without sparsification), I got a perplexity of around ~8.8 with the following parameters:

	•	max_length=16K
	•	stride=1, 2, 4, 8, 16K

I expected similar results for the sparse model, not NaN values.

**Environment**
I use [RunPod](https://www.runpod.io/) online env with A100-80GB-SXM *2

**To Reproduce**

	1.	Convert the Llama3-8B model using llm-compress to a sparse version.
	2.	Load the sparse model using **_SparseAutoModelForCausalLM_** (same process [here](https://github.com/vllm-project/llm-compressor/tree/main/examples/quantization_24_sparse_w4a16) ) and set up the environment to calculate perplexity.
	3.	Run perplexity calculation on WikiText2 dataset following Hugging Face’s [official perplexity guide](https://huggingface.co/docs/transformers/perplexity), but using the custom sparse model.
	4.	Observe the NaN perplexity values in the output.

**Errors**
Here’s the output I receive when running the perplexity calculation, see the attachment image. The perplexity of local Llama-8B model (load by _SparseAutoModelForCausalLM_ class) always be NaN value. Test with Llama-3B model (load by _AutoModelForCausalLM_ class) can successfully get ppl value.

### Sparse Llama 8B (load by _SparseAutoModelForCausalLM_ class) : ppl will be NaN
![LoadSparseLlama8BModel](https://github.com/user-attachments/assets/3fdfe7ea-a83f-4e8c-a22c-7d57b12171bc)
---
![PerplexityNaNOfSparseLlama8B](https://github.com/user-attachments/assets/98499e3f-b614-42be-9e14-c1483257227e)

### Load Online Llama 3B (load by _AutoModelForCausalLM_ class) : successfully get ppl value
![LoadLlama3BModel](https://github.com/user-attachments/assets/0fa42712-00cf-41f8-8d79-19db16cc4139)
---
![PerplexityOfLlama3B](https://github.com/user-attachments/assets/bf3270a9-ebe0-4863-8eb5-6759151eeca1)

**Additional context**
The same perplexity calculation process works perfectly when using the Hugging Face Llama-3.2-3B model without sparsification, which gives a perplexity value of ~8.8. I believe the issue lies in either the custom sparse model class or the integration with compressed-tensors. Maybe I miss some additional configuration/setting of Sparse model ? 🧐
Any guidance on this would be appreciated! 🥰

**Additional Question**
How to load the final quantization model (i.e the model be saved in _stage_quantization_ folder) correctly ?
I also interest of ppl of final quantization model, but I try load with _SparseAutoModelForCausalLM_ it can not be work 😢 
it shows some message mean : "... ... class not support ..."
So how to load the final quantization model correctly ? any documentation can be refer ? 🙏🏼

**Comment by robertgshaw2-neuralmagic** [comment]
Can you share the model?

**Comment by HengJayWang** [comment]
Hi @robertgshaw2-neuralmagic Robert, you were right to question this. I retested the original llama-7B Sparse conversion example from llm-compressor today, along with a simple `model.generate` test to check the model's text output. It turns out the model doesn’t seem to generate any correct outputs, and as expected, I couldn’t calculate the model’s perplexity under these circumstances.

## Load local Sparse Llama-7B model
![Sparse-Llama2-7B-load](https://github.com/user-attachments/assets/ced958fa-ba77-436f-b14f-47afbdac7a11)
### Test Model Output ([Ref](https://huggingface.co/docs/transformers/generation_strategies#greedy-search))
![Sparse-Llama2-7B-text-output](https://github.com/user-attachments/assets/d35d4d89-7776-4d8a-868a-ec2abaecedbf)
### Calculating Perplexity
![Sparse-Llama2-7B-ppl-calculating](https://github.com/user-attachments/assets/62255050-7dd2-419c-8b1b-f307f2975589)
### NaN Result
![Sparse-Llama2-7B-ppl-result](https://github.com/user-attachments/assets/fe2bf804-4998-4ec4-84a5-ed613fe37b7f)

I think the issue is now clearer. I believe the problem lies in how I load the local Sparse Model & Tokenizer. Does `llm-compressor` have any examples or documentation I can refer to? Any suggestions would be appreciated, thank you! 🥰

Also, I apologize for not providing the exact sparse model I used. After running it in the online RunPod environment, I didn’t download the model. However, this process should be easy to replicate. Here are the steps I followed for testing:

Step 1: Execute the official llama-7B sparse conversion example from [llm-compressor](https://github.com/vllm-project/llm-compressor/tree/main/examples/quantization_24_sparse_w4a16) :  run `python llama7b_sparse_w4a16.py`
Step 2: After about an hour, the sparse conversion finishes, and you’ll find the model saved in three stages in the output folder `output_llama7b_2:4_w4a16_channel` and I rename to `output_llama7b_2_4_w4a16_channel` for easy use.
Step 3: Load the stage_finetuing sparse model and Tokenizer in `output_llama7b_2_4_w4a16_channel/stage_finetuning`, and follow the [HuggingFace process](https://huggingface.co/docs/transformers/perplexity) to calculate perplexity"

##The Success Case with Llama3-3B online model
![Llama3-3B-load](https://github.com/user-attachments/assets/ba9677b6-ab4b-4da9-a9aa-1b6e10326101)
### Test Model Output
![Llama3-3B-text-output](https://github.com/user-attachments/assets/a08c7e7e-f128-4bf5-b1f4-558079dba219)
### Calculating Perplexity
![Llama3-3B-ppl-calculating](https://github.com/user-attachments/assets/4ff12b6c-5ca0-41de-bcc4-314536481e55)
### Result
![Llama3-3B-ppl-result](https://github.com/user-attachments/assets/8546b25c-baac-4d97-869b-9494cccf87a5)

## Summary
I want to correctly load the local sparse model and calculate its perplexity as an evaluation metric. However, it seems that I haven’t used the correct method to load the model (through the `SparseAutoModelForCausalLM` class) or the Tokenizer. If there are any documents or resources I can refer to, please let me know. Thanks! 🥰

And my testing jupyter notebook is in attatchment.
[Perplexity of  model.zip](https://github.com/user-attachments/files/17480086/Perplexity.of.model.zip)

**Comment by brian-dellabetta** [comment]
Hi @HengJayWang , so you've tried both `SparseAutoModelForCausalLM` from llmcompressor and `AutoModelForCausalLM` from huggingface transformers, and both result in NaNs?

**Comment by brian-dellabetta** [comment]
Hi @HengJayWang , I am able to run [the example you linked above](https://github.com/vllm-project/llm-compressor/blob/main/examples/quantization_24_sparse_w4a16/llama7b_sparse_w4a16.py) with `model_stub = "meta-llama/Meta-Llama-3-8B-Instruct"` and this pasted on the bottom of the script:

```python
import torch
from datasets import load_dataset
from tqdm import tqdm

def evaluate_perplexity(model, tokenizer):
    def _perplexity(nlls, n_samples, seqlen):
        return torch.exp(torch.stack(nlls).sum() / (n_samples * seqlen))

    # load and prepare dataset
    data = load_dataset("wikitext", "wikitext-2-raw-v1", split="test")
    data = tokenizer("\n\n".join(data["text"]), return_tensors="pt")
    data = data.input_ids.to(model.device)

    seqlen = 2048
    model = model.eval()
    n_samples = data.numel() // seqlen

    nlls = []

    with tqdm(range(n_samples), desc="Perplexity -") as progress_bar:
        for i in progress_bar:
            start_index = i * seqlen
            end_index = (i + 1) * seqlen
            batch = data[:, start_index:end_index].to(model.device)
            with torch.no_grad():
                logits = model(batch).logits
            shift_logits = logits[:, :-1, :].contiguous().float()
            shift_labels = data[:, start_index:end_index][:, 1:]
            loss_fct = torch.nn.CrossEntropyLoss()
            loss = loss_fct(
                shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1)
            )
            neg_log_likelihood = loss.float() * seqlen
            nlls.append(neg_log_likelihood)

            curr_ppl = _perplexity(nlls, i + 1, seqlen)
            progress_bar.set_description(f"Perplexity {curr_ppl:.3f}")

    ppl = _perplexity(nlls, n_samples, seqlen)

    return ppl.item()

ppl = evaluate_perplexity(model, tokenizer)

```

It takes ~45 minutes on an H100, but I get PPL vals in the 40s though, so poor recovery for this particular recipe. Any particular reason you would like to use this recipe?

**Comment by HengJayWang** [comment]
Hi Brian @brian-dellabetta 👋 ,

Thank you for sharing the test results. 🥰 I appreciate your effort in providing the code snippet. I noticed that there is [an updated version of the related 2of4_sparse code](https://github.com/vllm-project/llm-compressor/tree/main/examples/quantization_2of4_sparse_w4a16), and I will make sure to find some time to try out the code snippet you provided to see if it works as expected.

So based on your testing, did you use `SparseAutoModelForCausalLM` or `AutoModelForCausalLM` to load the sparse model and successfully calculate the perplexity in the end? I originally intended to use `SparseAutoModelForCausalLM `to load the model, because the model had already been processed by the SparseGPT algorithm and stored in sparse format, so it made sense to use `SparseAutoModelForCausalLM` to load the model. In any case, I will try the code snippet you tested. I hope llm-compressor can provide more detailed examples and documentation on sparse conversion in the future. Thank you. 🥰


Best regards, 
Jay Wang

**Comment by brian-dellabetta** [comment]
Hi @HengJayWang thanks for the feedback, will discuss with the team on sparsification examples/docs. Please let me know if you hit issues running any example or my snippet

Regarding `SpareAutoModelForCausalLM`, that has been deprecated because we have pushed those changes to transformers upstream directly. You should be fine to use `transformers.AutoModelForCausalLM` directly from here on out.

**Comment by brian-dellabetta** [comment]
One thing to note, that particular examples is set up to run quickly but you would want to update the training parameters to get good results. that is one reason why my PPL is so high. [This line in particular](https://github.com/vllm-project/llm-compressor/blob/main/examples/quantization_2of4_sparse_w4a16/llama7b_sparse_w4a16.py#L28)
